model/foreach_state.go

```diff
@@ -0,0 +1,113 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	val "github.com/serverlessworkflow/sdk-go/v2/validator"
+)
+
+func init() {
+	val.GetValidator().RegisterStructValidationCtx(ForEachStateStructLevelValidation, ForEachState{})
+}
+
+// ForEachModeType Specifies how iterations are to be performed (sequentially or in parallel)
+type ForEachModeType string
+
+const (
+	// ForEachModeTypeSequential specifies iterations should be done sequentially.
+	ForEachModeTypeSequential ForEachModeType = "sequential"
+	// ForEachModeTypeParallel specifies iterations should be done parallel.
+	ForEachModeTypeParallel ForEachModeType = "parallel"
+)
+
+// ForEachState used to execute actions for each element of a data set.
+type ForEachState struct {
+	BaseState
+	// Workflow expression selecting an array element of the states data
+	InputCollection string `json:"inputCollection" validate:"required"`
+	// Workflow expression specifying an array element of the states data to add the results of each iteration
+	OutputCollection string `json:"outputCollection,omitempty"`
+	// Name of the iteration parameter that can be referenced in actions/workflow. For each parallel iteration, this param should contain an unique element of the inputCollection array
+	IterationParam string `json:"iterationParam,omitempty"`
+	// Specifies how upper bound on how many iterations may run in parallel
+	BatchSize *intstr.IntOrString `json:"batchSize,omitempty"`
+	// Actions to be executed for each of the elements of inputCollection
+	Actions []Action `json:"actions,omitempty" validate:"required,min=1,dive"`
+	// State specific timeout
+	Timeouts *ForEachStateTimeout `json:"timeouts,omitempty"`
+	// Mode Specifies how iterations are to be performed (sequentially or in parallel)
+	// Defaults to parallel
+	Mode ForEachModeType `json:"mode,omitempty"`
+}
+
+type forEachStateForUnmarshal ForEachState
+
+func (f *ForEachState) UnmarshalJSON(data []byte) error {
+	v := forEachStateForUnmarshal{
+		Mode: ForEachModeTypeParallel,
+	}
+	err := json.Unmarshal(data, &v)
+	if err != nil {
+		return fmt.Errorf("forEachState value '%s' is not supported, it must be an object or string", string(data))
+	}
+
+	*f = ForEachState(v)
+	return nil
+}
+
+// ForEachStateStructLevelValidation custom validator for ForEachState
+func ForEachStateStructLevelValidation(_ context.Context, structLevel validator.StructLevel) {
+	stateObj := structLevel.Current().Interface().(ForEachState)
+
+	if stateObj.Mode != ForEachModeTypeParallel {
+		return
+	}
+
+	if stateObj.BatchSize == nil {
+		return
+	}
+
+	switch stateObj.BatchSize.Type {
+	case intstr.Int:
+		if stateObj.BatchSize.IntVal <= 0 {
+			structLevel.ReportError(reflect.ValueOf(stateObj.BatchSize), "BatchSize", "batchSize", "gt0", "")
+		}
+	case intstr.String:
+		v, err := strconv.Atoi(stateObj.BatchSize.StrVal)
+		if err != nil {
+			structLevel.ReportError(reflect.ValueOf(stateObj.BatchSize), "BatchSize", "batchSize", "gt0", err.Error())
```

Comment:
I guess this gt0 stands for greater than 0, it is fine, but, in the error it will be reported like:
`failed on the 'gt0' tag`

It is not clear why exactly it has failed.

Reply:
> I guess this gt0 stands for greater than 0, it is fine, but, in the error it will be reported like: `failed on the 'gt0' tag`
> 
> It is not clear why exactly it has failed.

It's the limit of `validator` library, we cann't report detail message other than tag name. Maybe we can fix it later.

Reply:
maybe we could try to use another thing as tag name.
maybe the field or subfield name.

In the future we could extend the validator to override these messages, I don't know how validator works, but I guess we could extend it, right?

Reply:
Yes, if we want to custom the error message we can do this like https://github.com/go-playground/validator/issues/559. I think we can create an issue for this.